[FEATURE] Future events

scribblecrumb opened the pull request:
<!-- Write BELOW The Headers and ABOVE The comments else it may not be viewable. -->
<!-- You can view CONTRIBUTING.md for a detailed description of the pull request process. -->
<!-- Be sure to name your PR something descriptive and succinct; include Bugfix: Feature: Enhancement: or Content: in the title to describe what type of PR it is. -->

## About The Pull Request
This adds FutureEvents, essentially a format block that can be added to any ShortEvent or Patrol that will trigger a future ShortEvent.

I've included documentation in this PR to explain it's use to writers, I think it serves as a good introduction to how this feature works.

A new class `FutureEvent` handles most things to do with this feature.  `prep_event` handles processing a ShortEvent or Patrol future_event block into a class object while also gathering the IDs of involved cats and deciding moon delay.

I've also added functions to handle the retrieval of a qualifying cat to `event_filters.py`

Within the `HandleShortEvents` class we handle triggering future events in `trigger_future_event`, a function that's called during the timeskip cycle in `events.py`.

The Clan's currently "queued" events are saved within a new `future_event.json` file within their Clan save folder.

You can also use the event editor to add these to short events!  However, the editor doesn't include the ability to add involved_cat constraints, because I'm sick of looking at the editor code and also I couldn't think of a way to do it well.  But, the vast majority of the time there's no reason to add constraints, it's far better to constrain the event pool rather than the involved cats.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage senior developers from merging your PR! -->

## Why This Is Good For ClanGen
Currently, this will be utilized for murder reveals.  Murder death events will now have a future event block attached to them that specifies a murder reveal to occur in the future.  The perk of this is that the "reveal cat", the cat who can confront a murderer about their actions, will be decided *when the murder event occurs*.  Meaning that we'll no longer have instances of cats confronting murderers about a murder that they could not feasibly have witnessed, a limitation of the prior murder reveal system.

<!-- If this is a bug fix, you can remove this section. -->
<!-- Please add a short description of why you think these changes would benefit the game. -->
<!-- If this PR was the result of discussion/testing on the discord, please add a link to the discord conversation here. -->
<!-- If you have multiple features that can stand on their own, or unrelated bugfixes, please create separate PRs for them. -->

## Proof of Testing
**Patrol Delay Testing**
Testing if patrols can trigger delayed events
![image](https://github.com/user-attachments/assets/19fedb7f-d98d-41b3-915e-947fb68c0639)
Added the gull feather accessory event as a delayed event for this patrol, then I ensured the patrol through the debug_ensure config. Patrol went off without a hitch.
![image](https://github.com/user-attachments/assets/29d11328-d2b6-42c3-a128-1cccab3a09e5)
The next moon, the delayed event appeared as expected.

**ShortEvent Delay Testing**
Testing if short events can trigger delayed events
![image](https://github.com/user-attachments/assets/19e18273-0196-4f5a-907f-714f2e254c3c)
I forced murders by lowering the chance to 1, so we'll have a heinous number of murders.  All murders now specify reveals as their delayed events.
![image](https://github.com/user-attachments/assets/2365c43b-aeb4-4082-b0ac-3ec4fb3d6419)
As you can see, though, murder reveals do occur the next moon

**Save loading**
![image](https://github.com/user-attachments/assets/bd2ff1a0-9a5e-404b-8cde-3cea37c05297)
Utilizing debug mode you can see that the save file was retrieved and it's data parsed correctly.

<!-- Include any screenshots, debugging steps, or links to beta testing threads here. At least one form of proof of testing is REQUIRED for all new content. You must be able to run the code locally before you PR it here. -->

## Changelog/Credits
- Murder reveals are now decided at the time of the murder, meaning that "reveal cats", the cat who may be involved in a murder reveal, will now be a cat who existed when the murder occurred.

<!-- Include any changes that should be made to the changelog of the game here or any changes to the credits file of the game. -->
<!-- This is just for easy access later for senior developers gathering this information; this process is not automated. -->

scribblecrumb committed:
DelayedEvent class

scribblecrumb committed:
shortevent handling of delayedevent

skeleton of handling delayedevents given to short events and a system to grab necessary info of cats who need to be involved going forward

scribblecrumb committed:
small adjustments to delayedevents

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into delayed-events

scribblecrumb committed:
move directories, cat collection

scribblecrumb committed:
event triggering

scribblecrumb committed:
excluded and allowed

scribblecrumb committed:
delayed events working

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into delayed-events

scribblecrumb committed:
put correct murder chance back

scribblecrumb committed:
integrate delayed_events into patrols

scribblecrumb committed:
save and load

scribblecrumb committed:
documentation and some format changes

- delayed events are now a list
- no more amount_of_events

scribblecrumb committed:
example and function rename

scribblecrumb committed:
typo

scribblecrumb committed:
add delay blocks to all murders

scribblecrumb committed:
warning addition

scribblecrumb committed:
fixing some bugs

scribblecrumb committed:
put range back

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into delayed-events

scribblecrumb committed:
remove unused import

scribblecrumb committed:
fix pylint?

scribblecrumb committed:
doc edits

scribblecrumb committed:
remove unused extension

scribblecrumb committed:
put murder chance back

scribblecrumb committed:
adding a tip

selkirks commented on docs/dev/writing/delayed.md:
```diff
@@ -0,0 +1,179 @@
+        "moon_delay": [1,1],
+        "involved_cats": {}
+        }
+    }
```
extra curly bracket here

selkirks commented on docs/dev/writing/delayed.md:
```diff
@@ -0,0 +1,179 @@
+This block can be added to the end of ShortEvent and Patrol formats.
+
+!!! note
+    The delayed event block is a *list*, this means that you could have multiple delayed event dictionaries contained within, each dictionary creating it's own triggered event.
```
should be "its own triggered event"

selkirks reviewed (dismissed):
generally easy to understand in my opinion! just pointed out a few typos. otherwise, looks great and is easy to digest and follow! the example you gave is great and easy to understand.

scribblecrumb committed:
doc updates

scribblecrumb committed:
doc update

SunlitFable commented:
looks great, very easy to digest and understand imo :]

might be worth noting what happens if a cat involved with the delayed event dies before it triggers (such as the murderer, when there's a queued reveal), but otherwise it's good!

scribblecrumb committed:
add buffer removal

scribblecrumb committed:
update docs

scribblecrumb commented:
Thanks Fable! Added some info about that.

scribblecrumb committed:
small wording change

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into delayed-events

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into delayed-events

j-gynn commented on docs/dev/writing/delayed.md:
```diff
@@ -0,0 +1,181 @@
+**"r_c": {...}**
+> In this line, we aren't carrying over any cat from the parent event. Instead, we're trying to find a new cat. We've decided this cat can only be a senior, so that constraint is added. A cat will be chosen from the currently living cats, excluding any cats already involved in this event. We could provide 0 constraints if we wanted any cat to have access, in that case we would just leave an empty ...
+
+The cat constraints that can be utilized here are the same as [ShortEvents](shortevents.md#r_cdictstr-various), with a few exclusions. The ***only*** constraints you *can* use are `age`, `status`, `skill`, `trait`, and `backstory`.
```
Does this link work when running the wiki locally? /genq

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into delayed-events

scribblecrumb committed:
renaming to future event

scribblecrumb committed:
updating documentation

scribblecrumb committed:
adding subtyping ignoring

scribblecrumb committed:
changing class funcs into a module

scribblecrumb committed:
more name switches

scribblecrumb committed:
transferring cat list filtering to event_filters.py

scribblecrumb committed:
ensure all cat params work with cat_for_event

scribblecrumb committed:
moving for loop

scribblecrumb committed:
include a target cat for relationship status checks

scribblecrumb committed:
formatting

scribblecrumb committed:
should be last bit of renaming

scribblecrumb committed:
fix save load

scribblecrumb committed:
fix pylint issue

scribblecrumb committed:
starting on adding editor tab

scribblecrumb committed:
removing target cat

i just don't think it's that useful to include relationship_status constraints here.  it complicates the event editor and future event format without providing much real utility.

scribblecrumb committed:
delayed event editor first pass

scribblecrumb committed:
future editor working

scribblecrumb committed:
adding future event functionalty

scribblecrumb committed:
fix mistype

scribblecrumb committed:
pylint fixes

scribblecrumb committed:
finally the editor is working

scribblecrumb committed:
adding future events through editor should be working smooth

scribblecrumb committed:
quick doc update

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into delayed-events

j-gynn commented on docs/dev/writing/future.md:
```diff
@@ -0,0 +1,184 @@
+
+Future events are special event blocks that can be added to the ShortEvent and Patrol outcome formats. They allow the writer to specify a ShortEvent that will be created in a given number of moons. Writers can even specify a larger pool of events for the future event to be chosen from.  Writers can also add multiple future events to a single event, meaning that one event could trigger multiple ev ...
+
+## Throughout this documentation we will refer to the FutureEvent format block as the "future event" and the event that is *chosen to eventually display* as the "future event". The event that initially creates the future event will be referred to as the "parent event".
```
I'm not going to lie, I have no idea what this is trying to tell me.

scribblecrumb replied:
Oop, that was smth I forgot to change with the name switch

j-gynn commented on docs/dev/writing/future.md:
````diff
@@ -0,0 +1,184 @@
+This specifies what cats can fill the roles within the future event. You can also use this to carry cats from the parent event into the future event. This is structured as a dictionary, with the **key** being the future event's cat role and the **value** being either a dictionary of constraints or a parent event's cat role.
+
+Example of how this looks in use, the parent event for this hypothetical event is a murder event:
+```json
````
```suggestion
```json lines
```
i *think* it's `json lines` to avoid IDEs throwing errors about it? may be wrong, double-check & resolve this if it doesn't stop it

scribblecrumb replied:
It doesn't stopping pycharm from throwing errors for me

j-gynn commented on docs/dev/writing/shortevents.md:
```diff
@@ -330,27 +332,28 @@ lowercase season names + "any"
+| "new_name"                                  | Ensure the cat takes on a clan-like name.                                                                                                                                                                                                                                                                                                                       ...
+| "old_name"                                  | Ensure the cat keeps their old (maybe loner or kittypet) name. Doesn't work for kittens or litters.                                                                                                                                                                                                                                                             ...
+| "kittypet"                                  | Gives the cat a kitty-pet type backstory. If "meeting" is also included, this tag will make the cat an kittypet outsider.                                                                                                                                                                                                                                       ...
+| "loner"                                     | Gives the cat a loner type backstory. If "meeting" is also included, this tag will make the cat an loner outsider.                                                                                                                                                                                                                                              ...
```
```suggestion
| "loner"                                     | Gives the cat a loner type backstory. If "meeting" is also included, this tag will make the cat a loner outsider.                                                                                                                                                                                                                                                  |
```

j-gynn commented on docs/dev/writing/shortevents.md:
```diff
@@ -330,27 +332,28 @@ lowercase season names + "any"
+| "can_birth"                                 | If same-sex breeding is OFF, make the cat female. Otherwise, 50 - 50 gender-roll.                                                                                                                                                                                                                                                                               ...
+| "new_name"                                  | Ensure the cat takes on a clan-like name.                                                                                                                                                                                                                                                                                                                       ...
+| "old_name"                                  | Ensure the cat keeps their old (maybe loner or kittypet) name. Doesn't work for kittens or litters.                                                                                                                                                                                                                                                             ...
+| "kittypet"                                  | Gives the cat a kitty-pet type backstory. If "meeting" is also included, this tag will make the cat an kittypet outsider.                                                                                                                                                                                                                                       ...
```
```suggestion
| "kittypet"                                  | Gives the cat a kitty-pet type backstory. If "meeting" is also included, this tag will make the cat a kittypet outsider.                                                                                                                                                                                                                                           |
```

j-gynn commented on docs/dev/writing/shortevents.md:
```diff
@@ -330,27 +332,28 @@ lowercase season names + "any"
+| "dead"                                      | This cat will be generated as a dead cat.                                                                                                                                                                                                                                                                                                                       ...
+| "male"                                      | Makes the cat male.                                                                                                                                                                                                                                                                                                                                             ...
+| "female"                                    | Makes the cat female.                                                                                                                                                                                                                                                                                                                                           ...
+| "can_birth"                                 | If same-sex breeding is OFF, make the cat female. Otherwise, 50 - 50 gender-roll.                                                                                                                                                                                                                                                                               ...
```
```suggestion
| "can_birth"                                 | If same-sex breeding is OFF, make the cat female. Otherwise, either male or female will be chosen at random.                                                                                                                                                                                                                                                                                   |
```

j-gynn commented on docs/dev/writing/shortevents.md:
```diff
@@ -330,27 +332,28 @@ lowercase season names + "any"
+| "old_name"                                  | Ensure the cat keeps their old (maybe loner or kittypet) name. Doesn't work for kittens or litters.                                                                                                                                                                                                                                                             ...
+| "kittypet"                                  | Gives the cat a kitty-pet type backstory. If "meeting" is also included, this tag will make the cat an kittypet outsider.                                                                                                                                                                                                                                       ...
+| "loner"                                     | Gives the cat a loner type backstory. If "meeting" is also included, this tag will make the cat an loner outsider.                                                                                                                                                                                                                                              ...
+| "rogue"                                     | Gives the cat a rogue type backstory. If "meeting" is also included, this tag will make the cat an rogue outsider.                                                                                                                                                                                                                                              ...
```
```suggestion
| "rogue"                                     | Gives the cat a rogue type backstory. If "meeting" is also included, this tag will make the cat a rogue outsider.                                                                                                                                                                                                                                                  |
```

j-gynn commented on docs/dev/writing/shortevents.md:
```diff
@@ -330,27 +332,28 @@ lowercase season names + "any"
+| "loner"                                     | Gives the cat a loner type backstory. If "meeting" is also included, this tag will make the cat an loner outsider.                                                                                                                                                                                                                                              ...
+| "rogue"                                     | Gives the cat a rogue type backstory. If "meeting" is also included, this tag will make the cat an rogue outsider.                                                                                                                                                                                                                                              ...
+| "clancat"                                   | Gives the cat a former-clancat type backstory. If "meeting" is also included, this tag will make the cat a former Clancat outsider.                                                                                                                                                                                                                             ...
+| "meeting"                                   | Make the cat an outsider (the patrol just met them, but they didn't join). That cat will never take a new clan-like name.                                                                                                                                                                                                                                       ...
```
```suggestion
| "meeting"                                   | Make the cat an outsider (the patrol just met them, but they didn't join). That cat will never take a new Clan-like name.                                                                                                                                                                                                                                           |
```

j-gynn commented on docs/dev/writing/shortevents.md:
```diff
@@ -330,27 +332,28 @@ lowercase season names + "any"
+| "rogue"                                     | Gives the cat a rogue type backstory. If "meeting" is also included, this tag will make the cat an rogue outsider.                                                                                                                                                                                                                                              ...
+| "clancat"                                   | Gives the cat a former-clancat type backstory. If "meeting" is also included, this tag will make the cat a former Clancat outsider.                                                                                                                                                                                                                             ...
+| "meeting"                                   | Make the cat an outsider (the patrol just met them, but they didn't join). That cat will never take a new clan-like name.                                                                                                                                                                                                                                       ...
+| "exists"                                    | Will attempt to find an existing outsider to utilize instead of creating a new one.  Keep in mind that this ONLY checks if the existing outsider matches indicated status, age, and backstory.  DO NOT use this tag in conjunction with creating litters or assigning parentage.                                                                                ...
```
```suggestion
| "exists"                                    | Will attempt to find an existing outsider to utilize instead of creating a new one. Keep in mind that this ONLY checks if the existing outsider matches indicated status, age, and backstory.  DO NOT use this tag in conjunction with creating litters or assigning parentage.                                                                                    |
```

j-gynn commented on docs/dev/writing/future.md:
Are age and status for FutureEvents determined at the time the parent event is rolled, or at the time the FutureEvent should trigger?

scribblecrumb replied:
At the time that it should trigger!  This is to avoid ending up with future events that no longer make sense for a cat's age/status.

j-gynn commented on resources/lang/en/screens/event_edit.en.json:
```diff
@@ -52,6 +52,10 @@
         "supply_adjust_info": "<b>adjust:</b> How will the supply change?",
         "supply_increase_info": "increase by:",
+        "future_info": "You can add any number of future events to this event. Future events will trigger after a specific number of moons have passed. You choose what event pool it chooses from, what cats will be involved in it, and how many moons it will take to trigger. Each event block will trigger a single event.",
+        "pool_info": "<b>pool:</b>If multiple subtypes are chosen, only events with all subtypes will be included in the pool. To add or remove event_ids, type them into the appropriate entry box and hit enter.",
```
```suggestion
        "pool_info": "<b>pool:</b> If multiple subtypes are chosen, only events with all subtypes will be included in the pool. To add or remove event_ids, type them into the appropriate entry box and hit enter.",
```

j-gynn commented on scripts/clan.py:
```diff
@@ -1094,6 +1098,59 @@ def save_disaster(self, clan=game.clan):
+                    "event_type": event.event_type,
+                    "pool": event.pool,
+                    "moon_delay": event.moon_delay,
+                    "involved_cats": event.involved_cats,
```
I know this is nitpicky, but it'd be preferable if this code was within the `FutureEvent` class, as this does directly relate to the relevant elements of `FutureEvent`. Then, we could just call `save_list.append(event.to_dict() for event in game.clan.future_events)` which is much cleaner.

j-gynn commented on scripts/events_module/event_filters.py:
```diff
@@ -380,3 +381,148 @@ def _check_cat_gender(cat, genders: list) -> bool:
+     cat. Keep in mind that this will search for a possible cat with the given relationship toward comparison cat.
+    """
+    # gather funcs to use
+    funct_dict = {
```
```suggestion
    func_dict = {
```
Nitpicky thing, but the shorthand term for "function" is "func". I've been driven mad enough by `cur_screen` (i am convinced it should be curr_screen, damn it!), so I'm gonna be annoying here and get it right before it gets locked in

j-gynn commented on scripts/events_module/event_filters.py:
```diff
@@ -380,3 +381,148 @@ def _check_cat_gender(cat, genders: list) -> bool:
+    return choice(allowed_cats).ID
+
+
+def _get_cats_with_age(cat_list: list, ages: list) -> list:
```
A very minor thing, but if we're doing these checks a lot, it makes sense to ask for `ages` to be input as a tuple, as they're much faster. Repeat for anything else in this general area that's a list of strings that we loop over or do `x in y` for.

j-gynn commented on scripts/events_module/short/handle_short_events.py:
```diff
@@ -1,9 +1,10 @@
-import random
+from random import choice, choices, randrange, sample, randint
```
Why this change?

scribblecrumb replied:
Truly, do not remember, must have been something I did way way back when I first started writing this 6 months ago.  I can switch it back if you'd prefer that?

j-gynn commented on scripts/screens/EventEditScreen.py:
```diff
@@ -435,6 +436,35 @@ def __init__(self, name=None):
 
+        self.future_element = {}
+        self.future_block_list: list = []
+        """The list of the currently loaded future blocks"""
```
```suggestion
        """The list of currently loaded future blocks"""
```

j-gynn commented on scripts/screens/EventEditScreen.py:
```diff
@@ -435,6 +436,35 @@ def __init__(self, name=None):
+        self.future_block_list: list = []
+        """The list of the currently loaded future blocks"""
+        self.selected_future_block_index: str = ""
+        """The list index for the future block currently viewed by the user. This is kept as a string due to it doubling
+             as the text for its button."""
```
Which is used more often, the int or the str form? /genq

scribblecrumb replied:
int is needed for the actual saved event info, but the string is used for all display and button purposes, thus str gets used more often.

j-gynn commented on scripts/screens/EventEditScreen.py:
```diff
@@ -695,10 +735,42 @@ def handle_event(self, event):
+                )
+            # FUTURE EVENT IDS
+            elif self.future_element["include_entry"].is_focused:
+                new_id = self.future_element["include_entry"].get_text()
+                block_info = self.get_selected_block_info()["pool"]["event_id"]
+                if new_id not in block_info:
+                    block_info.append(new_id)
+                else:
+                    block_info.remove(new_id)
+                text = ""
+                for id in block_info:
+                    text += f"'{id}'<br>"
+                self.future_element["include_display"].set_text(text)
+                self.future_element["include_entry"].set_text("")
+                self.editor_container.on_contained_elements_changed(
+                    self.future_element["include_display"]
+                )
+            elif self.future_element["exclude_entry"].is_focused:
+                new_id = self.future_element["exclude_entry"].get_text()
+                block_info = self.get_selected_block_info()["pool"]["excluded_event_id"]
+                if new_id not in block_info:
+                    block_info.append(new_id)
+                else:
+                    block_info.remove(new_id)
+                text = ""
+                for id in block_info:
+                    text += f"'{id}'<br>"
+                self.future_element["exclude_display"].set_text(text)
+                self.future_element["exclude_entry"].set_text("")
+                self.editor_container.on_contained_elements_changed(
+                    self.future_element["exclude_display"]
+                )
```
Could all this be moved into a function? Just because handle_event is already horrifyingly enormous in this screen.

j-gynn commented on scripts/screens/EventEditScreen.py:
```diff
@@ -2118,7 +2257,38 @@ def valid_supply(self) -> bool:
+                self.alert_text = "A Future Event block has no subtype, event_id, or excluded_event_id given. Event pool is too broad, you must use at least one of these constraints!"
+            elif block["moon_delay"][0] > block["moon_delay"][1]:
+                valid = False
+                self.alert_text = "A Future Event block has an invalid moon delay. The 2nd moon delay number should be equal to or larger than the 1st!"
```
```suggestion
                self.alert_text = "A Future Event block has an invalid moon delay. The second moon delay number should be equal to or larger than the first!"
```
Unless this is a britishism I'm not aware of, you should always write out numbers unless it's a date (I think the threshold for this is twenty, but I might be wrong there...)

j-gynn reviewed (changes requested):
ugh,,,, god, please let that be the last feature you want to add to the event editor,,, i don't think i have the strength for any more

scribblecrumb committed:
remove unnecessary paragraph

scribblecrumb committed:
Apply suggestions from code review

Co-authored-by: J Gynn <48025294+j-gynn@users.noreply.github.com>

scribblecrumb committed:
Apply suggestions from code review

Co-authored-by: J Gynn <48025294+j-gynn@users.noreply.github.com>

scribblecrumb committed:
adding to_dict func

also remove unused import

scribblecrumb committed:
Merge branch 'delayed-events' of https://github.com/scribblecrumb/clangen into delayed-events

scribblecrumb committed:
switching to tuples

scribblecrumb committed:
move to func
